### PR TITLE
Add initial machine learning pipeline

### PR DESCRIPTION
**Pull request by redshiftzero**
This PR adds an initial machine learning pipeline that takes the features in the database, trains a series of binary classifiers, evaluates how well each classifier performs, and then saves a bunch of relevant performance metrics in the database as well as pickling the trained model objects (for use in future scoring). The work in this PR corresponds to the latter half of this diagram from the `features` schema on: 
pipeline

A more complete description of our pipeline is described in `docs/pipeline.md` and a (very) brief description of how specialized classifiers might be integrated is stored in `CONTRIB.md`.

### CODE REVIEW

File: docs/Pipeline.md
Comment (psivesely) [inline docs/Pipeline.md]: This fails because the default Python in Xenial is 2.7. `./features.py` will read the shebang and choose the version appropriately. Be sure to fix the ones below too.

File: docs/Pipeline.md
Comment (psivesely) [inline docs/Pipeline.md]: Proposal to help the ML newbies like me understand what's going on here:

> `feature_scaling`: setting [feature scaling](https://en.wikipedia.org/wiki/Feature_scaling) will take the features and compute their [Standard score](https://en.wikipedia.org/wiki/Standard_score), re-scaling them to zero.... See also [scikit-learn's documentation](http://scikit-learn.org/stable/modules/preprocessing.html).

File: docs/Pipeline.md
Comment (psivesely) [inline docs/Pipeline.md]: Percent of what? I'll probably figure out later as I read on in the docs or source, but would be better if it was clarified here.

File: fpsd/attack.py
Comment (psivesely) [inline fpsd/attack.py:18]: The last code I wrote (see https://github.com/freedomofpress/fingerprint-securedrop/blob/fecb7fa21811e6f40d66fb097880b02807c0375e/fpsd/utils.py#L28), I wrote with the documentation style described in #53. I'm not absolutely set on one particular style, but we should be consistent. If you want to make other suggestions, do so in #53 and we can discuss, but if you also like the python-gnupg style, then you should make the appropriate changes here.

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py]: Don't see you using this dependency, and believe you meant to import the `interpolate` sub-package anyway.

File: fpsd/database.py
Comment (psivesely) [inline fpsd/database.py]: I implemented this functionality in a `Database` base class. You can get rid of this function, and replace line 52 below with `super().__init__(**kwargs)` .

File: fpsd/database.py
Comment (psivesely) [inline fpsd/database.py]: In Murrica we spell this labeled, but I just learned that spelling is normal for most other English speaking countries.

File: fpsd/attack.py
Comment (psivesely) [inline fpsd/attack.py]: The `load_closed_world` and `load_open_world` methods could easily be condensed into a single method by using format with query string. Then you could replace this block with the single line `df.db_closed_world()`.

File: fpsd/attack.py
Comment (psivesely) [inline fpsd/attack.py]: I would condense this into one line `... in generate_experiments(options)`.

File: fpsd/attack.py
Comment (psivesely) [inline fpsd/attack.py:55]: At first I though this was the multiplicative product, a note specifying it's the Cartesian product would help readability.

File: fpsd/attack.py
Comment (psivesely) [inline fpsd/attack.py]: This argument naming scheme is a little funky. Since there's no conflicts to worry about, why not `-c, --config`? The help string should make it clear.

File: fpsd/requirements/ml-requirements.txt
Comment (psivesely) [inline fpsd/requirements/ml-requirements.txt]: Can you put the top-level imports only in the `fpsd/requirements/requirements.in` file?

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py]: I think setting `self.n_cores = multiprocessing.cpu_count()` in `__init__` is probably better than assuming there's 2.

File: fpsd/attack.yaml
Comment (psivesely) [inline fpsd/attack.yaml]: What editor are you using that does this weird no newline at end of line thing?

File: fpsd/attack.yaml
Comment (psivesely) [inline fpsd/attack.yaml:7]: It might be fun to make this a hyperparameter to play with as well. Would help in threat modeling. Though it is computationally costly. So it could just be set to `[0.20]` initially, and then if we ever want to investigate later, we can add some other values.

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py]: You can simplify this to `class Experiment:`.

File: fpsd/requirements/ml-requirements.txt
Comment (psivesely) [inline fpsd/requirements/ml-requirements.txt]: This should not be here. It's installed by default except in Travis by an Ansible task.

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py:112]: Do it! (#64)

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py:172]: Why not just unpack the `self.hyperparameters` wherever you can below (i.e., when the key names for the values you wish to pass line up with the corresponding arguments)?

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py:109]: Did you consider using `sklearn.cross_validation.ShuffleSplit` here to address #61? Thoughts either way?

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py:159]: We should talk about the suggestions in http://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations. I'm personally in favor of pinning our version of `sklearn` in the `.in` file, and saving all the metadata they recommend along with the pickle modeled. If a new version of `sklearn` comes out with some fantastic features we want to use, it won't be the end of the world to retrain all our old models.

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py]: Can you keep the var as `x_test` in this fn. for consistency?

File: docs/Pipeline.md
Comment (psivesely) [inline docs/Pipeline.md:85]: I think in WTF-PAD they correctly point out that precision and recall are more important measures than TPR and FPR. They considered F<sub>1</sub> to be the most important metric, but I personally believe that F<sub>0.5</sub> is closer to the most informative single metric we can look at given our specific problem. There is a function `sklearn.metrics.fbeta_score` that will compute F<sub>B</sub>.

File: fpsd/evaluation.py
Comment (psivesely) [inline fpsd/evaluation.py]: Can you drop the weird period? I know it has no effect, but...

File: fpsd/evaluation.py
Comment (psivesely) [inline fpsd/evaluation.py:58]: I don't understand this comment or what you're doing here. Can you please explain?

File: fpsd/evaluation.py
Comment (psivesely) [inline fpsd/evaluation.py:44]: You could condense this function significantly by using the one-liner

```
statistics.mean(map(lambda x: x[metric], metrics_list))
```

to compute the mean of a given metric over the folds.

File: fpsd/evaluation.py
Comment (psivesely) [inline fpsd/evaluation.py]: Why are you creating blank entries for these?

File: fpsd/evaluation.py
Comment (psivesely) [inline fpsd/evaluation.py:76]: No idea what this function is doing or why. Needs at least a docstring, maybe a comment in `get_metrics` as well. Also, can you move this function right below `get_metrics` since they are related and the two functions above this one are not related to `get_metrics`?

File: fpsd/classify.py
Comment (psivesely) [inline fpsd/classify.py]: There's a method that does this, `self.__dict__`. I would just rename the object attributes to match `sklearn`'s. More clear that way anyway.

File: fpsd/database.py
Comment (psivesely) [inline fpsd/database.py]: `class DatasetLoader(Database)`--classes below too.

File: fpsd/database.py
Comment (psivesely) [inline fpsd/database.py]: `with self.safe_session() as session:`--and below too.

File: fpsd/evaluation.py
Comment (psivesely) [inline fpsd/evaluation.py]: This attribute is cool. If we can identify features that are consistently found not to be useful for our top classifiers, we could use this information to help us implement the feature selection stage described in #63.

**Review by psivesely — DISMISSED** [review]
After changes and a rebase are made, I'll take another look. I'll do my part to help us iterate faster than happened with the features branch.

File: docs/Pipeline.md
Comment (redshiftzero) [reply docs/Pipeline.md]: Word - changed. Also changed attack.py below

File: docs/Pipeline.md
Comment (redshiftzero) [reply docs/Pipeline.md]: Added these links!

File: docs/Pipeline.md
Comment (redshiftzero) [reply docs/Pipeline.md]: Percentage of the testing set, explicitly added

File: docs/Pipeline.md
Comment (redshiftzero) [reply docs/Pipeline.md:85]: It’s true that precision, recall, and F1 score are better metrics, but for this first pass we are using AUC since it’s independent of class balance in the testing set (see Figure 5 in [this paper](http://people.inf.elte.hu/kiss/12dwhdm/roc.pdf)). I filed #62 to implement precision, recall, F1 storage. Not following why F_{0.5} would be a better metric?

File: fpsd/attack.py
Comment (redshiftzero) [reply fpsd/attack.py]: OK, condensed this into one function

File: fpsd/attack.py
Comment (redshiftzero) [reply fpsd/attack.py]: Fixed

File: fpsd/attack.py
Comment (redshiftzero) [reply fpsd/attack.py:55]: Done

File: fpsd/attack.py
Comment (redshiftzero) [reply fpsd/attack.py]: OK changed to config

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py]: Removed this.

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py]: Fixed

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py]: Good call, added this

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py]: Fixed

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py:109]: Not sure how this would address #61, but it’s not the best cross validation method to use here as it we want to make sure we are preserving the class balance as well as the fact that we’d have to specify the size of the training set, which `cross_validation.StratifiedKFold` does automatically by our specifying `k`

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py:112]: It was recommended that I make smaller PRs ^_^ this is why I created issue #64 and then submitted this PR before implementing this option. After this is merged, I will implement this in a separate PR. Trying to do the smaller PR thing to make incremental changes that are easier to review, but this means there are some TODO placeholders in the code that we will have to accept for now

File: fpsd/database.py
Comment (redshiftzero) [reply fpsd/database.py]: Done

File: fpsd/database.py
Comment (redshiftzero) [reply fpsd/database.py]: Done

File: fpsd/database.py
Comment (redshiftzero) [reply fpsd/database.py]: when in Rome I guess  ¯_(ツ)_/¯. Fixed

File: fpsd/database.py
Comment (redshiftzero) [reply fpsd/database.py]: Done

File: fpsd/requirements/ml-requirements.txt
Comment (redshiftzero) [reply fpsd/requirements/ml-requirements.txt]: Removed

File: fpsd/requirements/ml-requirements.txt
Comment (redshiftzero) [reply fpsd/requirements/ml-requirements.txt]: Done

File: fpsd/attack.yaml
Comment (redshiftzero) [reply fpsd/attack.yaml:7]: Yep we can make this take a list at a later point, updated #64 issue to keep this suggestion

File: fpsd/attack.yaml
Comment (redshiftzero) [reply fpsd/attack.yaml]: I used Sublime Text to write this ¯_(ツ)_/¯.  Added newline

File: fpsd/evaluation.py
Comment (redshiftzero) [reply fpsd/evaluation.py]: Placeholders to be implemented in #62, added inline TODO

File: fpsd/evaluation.py
Comment (redshiftzero) [reply fpsd/evaluation.py]: Done

File: fpsd/evaluation.py
Comment (redshiftzero) [reply fpsd/evaluation.py:58]: The TPR and FPR output from each individual fold need to be averaged (not _quite_ that simple since they are arrays that need to be interpolated onto a common threshold grid and then averaged) and saved. This is a TODO (added inline TODO) and is issue #65

File: fpsd/evaluation.py
Comment (redshiftzero) [reply fpsd/evaluation.py]: :+1:

File: fpsd/evaluation.py
Comment (redshiftzero) [reply fpsd/evaluation.py:76]: Added docstring, moved up in the file

File: fpsd/attack.py
Comment (redshiftzero) [reply fpsd/attack.py:18]: I was using [Google style docstrings](http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#example-google), but don’t have strong feelings one way or the other so the docstring format you like there is fine w/me and from now on I will follow this format. However, most of the existing docstrings in this branch I wrote before that issue was filed so would rather not rewrite the Google style docstrings at this stage in the interests of time.

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py:172]: Here is my concern: I think this is already difficult to understand for people not very familiar with scikit-learn classifiers and their hyperparameters. I'm worried that by using unpacking here this will become even more difficult to understand - it reduces the number of lines of code, but also negatively impacts readability. Thoughts?

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py:159]: Lots of good comments in the link you share here. Some of the metadata in there is being saved (e.g. cross-validation scores) in the database, however I have now pinned version 0.17.0 of scikit-learn in requirements.in.

File: fpsd/evaluation.py
Comment (redshiftzero) [reply fpsd/evaluation.py:44]: Hm maybe I’m being dense right now but this doesn’t seem to work for me due to the scope here (I get `’metric’ is not defined`)?

File: fpsd/classify.py
Comment (redshiftzero) [reply fpsd/classify.py]: Done in 5a474343

**Comment by redshiftzero** [comment]
OK comments addressed, ansible-ification of the creation of the models schema and tables is done, Travis builds are passing, and it's rebased on current master. Should be good to go 🌞

**Comment by conorsch** [comment]
What a review process this has been! Thanks for your patience here, @redshiftzero. Given the frequent back-and-forth here, I'm inclined to merge, and we can bite off smaller hunks to discuss in discrete issues going forward.

**Comment by redshiftzero** [comment]
👍  sounds good - any other outstanding problems we can make issues for and address in smaller PRs

**Review by conorsch — APPROVED** [review]
Changes requested during review have been implemented. :shipit:

File: docs/Pipeline.md
Comment (psivesely) [reply docs/Pipeline.md:85]: Let's discuss in-person.

File: fpsd/attack.py
Comment (psivesely) [reply fpsd/attack.py:18]: Sgtm.

File: fpsd/classify.py
Comment (psivesely) [reply fpsd/classify.py:109]: Precisely because it allows us to to balance representation as we like, which is something roughly suggested in #61.

>ShuffleSplit is thus a good alternative to KFold cross validation that allows a finer control on the number of iterations and the proportion of samples on each side of the train / test split.

--http://scikit-learn.org/stable/modules/cross_validation.html#random-permutations-cross-validation-a-k-a-shuffle-split

File: fpsd/classify.py
Comment (psivesely) [reply fpsd/classify.py:172]: Your call/ @conorsch merged this without me already, so it's in.

File: fpsd/evaluation.py
Comment (psivesely) [reply fpsd/evaluation.py:44]: Might need to change a variable name... would much of the last suite starting at line 63 I believe.

**Pull request merged**
